File: C99/백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdint.h>

/**
 * @description DFS
 * @performance 메모리: 1,112 KB, 동작시간: 0 ms
 * @author java08
 */

// 20병으로 갈 수 있는 최대 거리: 1000미터
// DFS로 각 편의점 지점을 거쳐가며 목적지까지 도착할 수 있는지 확인

static inline int code(int x, int y)
{
    return (x << 16) | (y & 0xFFFF);
}

static inline void decode(int target, int *x, int *y)
{
    *x = target >> 16;
    *y = (target << 16) >> 16;
}

static inline int get_dist(int a, int b)
{
    return a > b ? a - b : b - a;
}

int main(void)
{
    int t;
    if (scanf("%d", &t) != 1)
        return 1;
    while (t-- > 0)
    {
        int n;
        if (scanf("%d", &n) != 1)
            return 1;
        int node[101];
        int x, y;
        if (scanf("%d %d", &x, &y) != 2)
            return 1;
        node[0] = code(x, y);
        for (int i = 1; i <= n; i++)
        {
            if (scanf("%d %d", &x, &y) != 2)
                return 1;
            node[i] = code(x, y);
        }
        if (scanf("%d %d", &x, &y) != 2)
            return 1;
        bool visited[101] = {0};
        int stack[101];
        int top = 0;
        stack[top++] = 0;
        visited[0] = true;

        bool flag = false;
        while (top)
        {
            int curIdx = stack[--top];
            int curX, curY;
            decode(node[curIdx], &curX, &curY);
            int distG = get_dist(curX, x) + get_dist(curY, y);
            if (distG <= 1000)
            {
                flag = true;
                break;
            }

            for (int i = 1; i <= n; i++)
            {
                if (visited[i])
                    continue;
                int nextX, nextY;
                decode(node[i], &nextX, &nextY);
                int dist = get_dist(curX, nextX) + get_dist(curY, nextY);
                if (dist <= 1000)
                {
                    visited[i] = true;
                    stack[top++] = i;
                }
            }
        }
        if (flag)
            printf("happy\n");
        else
            printf("sad\n");
    }
    return 0;
}

File: Java/백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.java
import java.io.*;
import java.util.*;

/**
 * @description BFS
 * @performance 메모리: ? KB, 동작시간: ? ms
 * @author python98
 */
public class Main {

	private static int nextInt() throws IOException{
		int s=1, n, c;
		while((n = System.in.read())<=32);
		if(n=='-') {
			s = -1;
			n = System.in.read() & 15;
		} else {
			n &= 15;
		}
		while((c = System.in.read())>32) {
			n = (n<<3) + (n<<1) + (c&15);
		}
		return s*n;
	}

	private static int getDist(int a, int b){
		return a>b?a-b:b-a;
	}
	public static void main(String[] args) throws IOException {
		StringBuilder sb = new StringBuilder();
		for(int t = nextInt() ; t > 0 ; t--) {
			int[] node = new int[101];
			int n = nextInt();
			int x = nextInt();
			int y = nextInt();
			node[0] = (x<<16) | (y&0xFFFF);
			for(int i = 1 ; i <= n ; i++){
				node[i] = (nextInt()<<16) | (nextInt()&0xFFFF);
			}
			x = nextInt();
			y = nextInt();

			int[] stack = new int[101];
			int top = 0;
			stack[top++] = 0;
			boolean[] visited = new boolean[101];
			visited[0] = true;
			boolean flag = false;
			while(top>0) {
				int curIdx = stack[--top];
				int curX = node[curIdx]>>16;
				int curY = (node[curIdx]<<16)>>16;
				int dist = getDist(curX,x)+getDist(curY,y);
				if(dist<=1000) {
					flag = true;
					break;
				}
				for(int i = 1 ; i <= n ; i++) {
					if(visited[i]) continue;
					int nextX = node[i]>>16;
					int nextY = (node[i]<<16)>>16;
					dist = getDist(curX,nextX)+getDist(curY,nextY);
					if(dist<=1000) {
						visited[i] = true;
						stack[top++] = i;
					}
				}
			}
			if(flag) {
				sb.append("happy\n");
			} else {
				sb.append("sad\n");
			}
		}
		System.out.print(sb);
	}
}